Core/HLE/HLE.cpp

```diff
@@ -354,6 +403,136 @@ bool hleIsKernelMode() {
 	return latestSyscall && (latestSyscall->flags & HLE_KERNEL_SYSCALL) != 0;
 }
 
+void hleEnqueueCall(u32 func, int argc, const u32 *argv, PSPAction *afterAction) {
+	std::vector<u32> args;
+	args.resize(argc);
+	memcpy(args.data(), argv, argc * sizeof(u32));
+
+	enqueuedMipsCalls.push_back({ func, afterAction, args });
+
+	hleAfterSyscall |= HLE_AFTER_QUEUED_CALLS;
+}
+
+void hleFlushCalls() {
+	u32 &sp = currentMIPS->r[MIPS_REG_SP];
+	PSPPointer<HLEMipsCallStack> stackData;
+	VERBOSE_LOG(HLE, "Flushing %d HLE mips calls from %s, sp=%08x", (int)enqueuedMipsCalls.size(), latestSyscall ? latestSyscall->name : "?", sp);
+
+	// First, we'll add a marker for the final return.
+	sp -= sizeof(HLEMipsCallStack);
+	stackData.ptr = sp;
+	stackData->nextOff = 0xFFFFFFFF;
+	stackData->ra = currentMIPS->pc;
+	stackData->v0 = currentMIPS->r[MIPS_REG_V0];
+	stackData->v1 = currentMIPS->r[MIPS_REG_V1];
+
+	// Now we'll set up the first in the chain.
+	currentMIPS->pc = enqueuedMipsCalls[0].func;
+	currentMIPS->r[MIPS_REG_RA] = HLEMipsCallReturnAddress();
+	for (int i = 0; i < (int)enqueuedMipsCalls[0].args.size(); i++) {
+		currentMIPS->r[MIPS_REG_A0 + i] = enqueuedMipsCalls[0].args[i];
+	}
+
+	// For stack info, process the first enqueued call last, so we run it first.
+	// We don't actually need to store 0's args, but keep it consistent.
+	for (int i = (int)enqueuedMipsCalls.size() - 1; i >= 0; --i) {
+		auto &info = enqueuedMipsCalls[i];
+		u32 stackRequired = (int)info.args.size() * sizeof(u32) + sizeof(HLEMipsCallStack);
+		u32 stackAligned = ((stackRequired + 15) / 16) * 16;
```

Comment:
I'd write this as `u32 stackAligned = (stackRequired + 0xF) & ~0xF, but doesn't really matter.